frame/vesting/src/lib.rs

```diff
@@ -312,105 +368,384 @@ pub mod pallet {
+		/// - `starting_block`: `MAX(schedule1.starting_block, scheduled2.starting_block, current_block)`.
+		/// - `ending_block`: `MAX(schedule1.ending_block, schedule2.ending_block)`.
+		/// - `locked`: `schedule1.locked_at(current_block) + schedule2.locked_at(current_block)`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schedule1_index`: index of the first schedule to merge.
+		/// - `schedule2_index`: index of the second schedule to merge.
+		#[pallet::weight(
+			T::WeightInfo::not_unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get())
+			.max(T::WeightInfo::unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get()))
+		)]
+		pub fn merge_schedules(
+			origin: OriginFor<T>,
+			schedule1_index: u32,
+			schedule2_index: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			if schedule1_index == schedule2_index {
+				return Ok(());
+			};
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+			let schedules = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+
+			// The schedule index is based off of the schedule ordering prior to filtering out any
+			// schedules that may be ending at this block.
+			let schedule1 =
+				*schedules.get(schedule1_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let schedule2 =
+				*schedules.get(schedule2_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let merge_action = VestingAction::Merge(schedule1_index, schedule2_index);
+
+			// The length of `schedules` decreases by 2 here since we filter out 2 schedules.
+			// Thus we know below that we can insert the new merged schedule without error (assuming
+			// the its initial state was valid).
+			let (mut schedules, mut locked_now) =
+				Self::report_schedule_updates(schedules.to_vec(), merge_action);
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			let now = <frame_system::Pallet<T>>::block_number();
+			if let Some(new_schedule) = Self::merge_vesting_info(now, schedule1, schedule2)? {
+				// Merging created a new schedule; so now we need to add it to the
+				// accounts vesting schedule collection.
+				schedules.push(new_schedule);
+				// (We use `locked_at` in case this is a schedule that started in the past.)
+				let new_schedule_locked = new_schedule.locked_at::<T::BlockNumberToBalance>(now);
+				// update the locked amount to reflect the schedule we just added,
+				locked_now = locked_now.saturating_add(new_schedule_locked);
+				// and deposit an event to show the new, new, merged schedule.
+				Self::deposit_event(Event::<T>::MergedScheduleAdded(
+					new_schedule.locked(),
+					new_schedule.per_block(),
+					new_schedule.starting_block(),
+				));
+			} // In the None case there was no new schedule to account for.
```

Comment:
Do we want an event in the `None` case?

Reply:
in general we want to minimize the footprint of events. you must evaluate whether this is significant to track

Reply:
probably fine without

Reply:
Tbh I think we could forego events in both cases since the logic could be replicated off-chain and simply having an extrinsic success would be enough. 

That being said we don't have the tooling yet to calc that logic (maybe I can add a polkadot-js derive) so you can go back to arguing the event(s) is/are helpful

Reply:
i agree, in general probably we dont need an event here. Also would be easy to add later if we were asked for it

Reply:
Merge event removed c3cae6b05d5059903bf2d3ce04367bf65309bed8